ts/session/apis/snode_api/SnodeRequestTypes.ts

```diff
@@ -1,71 +1,90 @@
-    namespace: number;
-  } & RetrieveAlwaysNeeded &
-    RetrieveMaxCountSize;
-};
+type WithMaxSize = { max_size?: number };
+export type WithShortenOrExtend = { shortenOrExtend: 'shorten' | 'extend' | '' };
 
-/** Those namespaces do not require to be authenticated for storing messages.
- *  -> 0 is used for our swarm, and anyone needs to be able to send message to us.
- *  -> -10 is used for legacy closed group and we do not have authentication for them yet (but we will with the new closed groups)
- *  -> others are currently unused
- *
+/**
+ * This is the base sub request class that every other type of request has to extend.
  */
-// type UnauthenticatedStoreNamespaces = -30 | -20 | -10 | 0 | 10 | 20 | 30;
-
-export type RetrieveLegacyClosedGroupSubRequestType = {
-  method: 'retrieve';
-  params: {
-    namespace: SnodeNamespaces.ClosedGroupMessage; // legacy closed groups retrieve are not authenticated because the clients do not have a shared key
-  } & RetrieveAlwaysNeeded &
-    RetrieveMaxCountSize;
-};
+abstract class SnodeAPISubRequest {
+  public abstract method: string;
 
-export type RetrieveSubKeySubRequestType = {
-  method: 'retrieve';
-  params: {
-    subkey: string; // 32-byte hex encoded string
-    signature: string;
-    namespace: number;
-  } & RetrieveAlwaysNeeded &
-    RetrieveMaxCountSize;
-};
+  public abstract loggingId(): string;
+  public abstract getDestination(): PubkeyType | GroupPubkeyType | '<none>';
+  /**
+   * When batch sending an array of requests, we will sort them by this number (the smallest will be put in front and the largest at the end).
+   * This is needed for sending and polling for 03-group keys for instance.
+   */
 
-export type RetrieveSubRequestType =
-  | RetrieveLegacyClosedGroupSubRequestType
-  | RetrievePubkeySubRequestType
-  | RetrieveSubKeySubRequestType
-  | UpdateExpiryOnNodeSubRequest;
+  public requestOrder() {
+    return 0;
+  }
+}
 
 /**
- * OXEND_REQUESTS
+ * Retrieve for legacy was not authenticated
  */
-export type OnsResolveSubRequest = {
-  method: 'oxend_request';
-  params: {
-    endpoint: 'ons_resolve';
-    params: {
-      type: 0;
-      name_hash: string; // base64EncodedNameHash
+export class RetrieveLegacyClosedGroupSubRequest extends SnodeAPISubRequest {
+  method = 'retrieve' as const;
+  public readonly legacyGroupPk: PubkeyType;
+  public readonly last_hash: string;
+  public readonly max_size: number | undefined;
+  public readonly namespace = SnodeNamespaces.LegacyClosedGroup;
+
+  constructor({
+    last_hash,
+    legacyGroupPk,
+    max_size,
+  }: WithMaxSize & { last_hash: string; legacyGroupPk: PubkeyType }) {
+    super();
+    this.legacyGroupPk = legacyGroupPk;
+    this.last_hash = last_hash;
+    this.max_size = max_size;
+  }
+
+  public build() {
+    return {
+      method: this.method,
+      params: {
+        namespace: this.namespace, // legacy closed groups retrieve are not authenticated because the clients do not have a shared key
```

Comment:
idk, I specifically don't want to move it as it is part of making sure no one adds the signature field here
```ts
      params: {
        namespace: this.namespace,
        pubkey: this.legacyGroupPk,
        last_hash: this.last_hash,
        max_size: this.max_size,
        // legacy closed groups retrieve are not authenticated because the clients do not have a shared key
        // if we give a timestamp, a signature will be requested by the snode so this request for legacy does not take a timestamp
      },
```
so I think I'll keep it there. Anyway, legacy groups are deprecated soon :tm: